File: src/api/endpoints/DefaultApiConfig.ts
import ApiConfig from "../ApiConfig";
import AccessTokenProvider from "../auth/AccessTokenProvider";

/**
 * Normalizes API configuration and prevents ambiguous authentication choices.
 */
export default class DefaultApiConfig {
  public readonly baseUrl: string;
  public readonly accessToken?: string;
  public readonly refreshToken?: string;
  public readonly cookieAuth: boolean;
  public readonly accessTokenProvider?: AccessTokenProvider;

  constructor(config: ApiConfig) {
    const hasLegacyAuthentication =
      config.accessToken !== undefined || config.refreshToken !== undefined || config.cookieAuth === true;

    if (config.accessTokenProvider !== undefined && hasLegacyAuthentication) {
      throw new Error(
        "accessTokenProvider cannot be combined with accessToken, refreshToken, or cookieAuth authentication.",
      );
    }

    this.baseUrl = config.baseUrl ?? "https://api.medrunner.space";
    this.accessToken = config.accessToken;
    this.refreshToken = config.refreshToken;
    this.cookieAuth = config.cookieAuth ?? false;
    this.accessTokenProvider = config.accessTokenProvider;
  }
}

File: src-dotnet/api/DefaultApiConfig.cs
namespace MedrunnerApiClient.Api
{
    /// <summary>
    /// Represents API config with defaults applied.
    /// </summary>
    public class DefaultApiConfig : ApiConfig
    {
        public DefaultApiConfig(ApiConfig config)
        {
            BaseUrl = string.IsNullOrEmpty(config.BaseUrl) ? "https://api.medrunner.space" : config.BaseUrl;
            AccessToken = config.AccessToken;
            RefreshToken = config.RefreshToken;
            CookieAuth = config.CookieAuth;
        }
    }
}
